libyul/optimiser/DataFlowAnalyzer.cpp

```diff
@@ -62,26 +62,26 @@ void DataFlowAnalyzer::operator()(ExpressionStatement& _statement)
 	if (auto vars = isSimpleStore(StoreLoadLocation::Storage, _statement))
 	{
 		ASTModifier::operator()(_statement);
-		set<YulString> keysToErase;
-		for (auto const& item: m_storage.values)
+		auto it = m_storage.values.begin();
+		while (it != m_storage.values.end())
 			if (!(
-				m_knowledgeBase.knownToBeDifferent(vars->first, item.first) ||
-				m_knowledgeBase.knownToBeEqual(vars->second, item.second)
+				m_knowledgeBase.knownToBeDifferent(vars->first, it->first) ||
+				m_knowledgeBase.knownToBeEqual(vars->second, it->second)
 			))
-				keysToErase.insert(item.first);
-		for (YulString const& key: keysToErase)
-			m_storage.eraseKey(key);
+				it = m_storage.values.erase(it);
+			else
+				++it;
 		m_storage.set(vars->first, vars->second);
 	}
 	else if (auto vars = isSimpleStore(StoreLoadLocation::Memory, _statement))
 	{
 		ASTModifier::operator()(_statement);
-		set<YulString> keysToErase;
-		for (auto const& item: m_memory.values)
-			if (!m_knowledgeBase.knownToBeDifferentByAtLeast32(vars->first, item.first))
-				keysToErase.insert(item.first);
-		for (YulString const& key: keysToErase)
-			m_memory.eraseKey(key);
+		auto it = m_memory.values.begin();
+		while (it != m_memory.values.end())
+			if (!m_knowledgeBase.knownToBeDifferentByAtLeast32(vars->first, it->first))
+				it = m_memory.values.erase(it);
+			else
+				++it;
```

Comment:
Actually, we could replace these with ``std::erase_if``, that's even specified to do the same for maps... https://en.cppreference.com/w/cpp/container/map/erase_if

Reply:
Ah, or damn, no, we can't - those are C++20 :-).

Reply:
We should probably come up with a nice way to add functions like that without running into ambiguity problems later... i.e. something like a ``cxx20`` namespace which we never import, but always only use fully qualified... and which can then later just be dropped... or something like that... and there's the question if we do it ourselves whereever we want or if we want to pull in some header-only libraries that already does it extensively... Not sure how good e.g. https://abseil.io/ or similar projects are...

Reply:
yep sounds good. Either boost or our own implementation.